### DISCUSSION TITLE
Is there an MikroORM way to check if a table is empty?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2547] [Q&A] [answered]
I can always use the following command to check if a table has at least one row, but is there a built-in way to check if a table is empty?

```sql
SELECT 1 FROM schema.table LIMIT 1;
```

**Update**

Is the following code my only/best option to check if the table is empty?

```ts
const qb = this.orm.em.createQueryBuilder(Entity)
const isEmpty = await qb.select('1 AS EMPTY').limit(1).withSchema(schema)
```

**Comment by B4nan** [comment]
> I can always use the following command to check if a table has at least one row, but is there a built-in way to check if a table is empty?

That seems wrong to me, table without any rows will also return 1, or not? You are checking existence of the table, not its values. 

edit: ok I though you were talking about information schema, got confused by `schema.table`.

> Is the following code my only/best option to check if the table is empty?

Still looks the same to me, if you want to know if table is empty, you need to check number if its rows, I would expect that query to give you same results regardless of number of items in the table - and if something, I'd expect this to give you EMPTY=1 in the exact opposite case - when there are some items. 

I'd say you want a count query.

**Reply by tukusejssirs** [reply]
> That seems wrong to me, table without any rows will also return 1, or not? You are checking existence of the table, not its values.

No. If the table has at least one row, it returns `1`, otherwise it returns nothing (an empty string). The command works, I use it in production.

```sql
> create table a (a INT);
CREATE TABLE

> select 1 as isEmpty from a limit 1;
 isempty 
---------
(0 rows)

> insert into a (a) values (3);
INSERT 0 1

> select 1 as isEmpty from a limit 1;
 isempty 
---------
       1
(1 row)
```

> I'd say you want a count query.

It’d take too long. I have a table (per machine) with a record (row) per second (~86400 rows a day; I want to get rid of this table, but for now I need to keep it for backward compatibility).

Anyway, the SQL command above works, but the MikroORM version does not; it throws the following error:

```ts
[Nest] 519983  - 17/12/2021, 10:34:18   ERROR [ExceptionsHandler] You cannot merge entity 'Entity' without identifier!
ValidationError: You cannot merge entity 'Entity' without identifier!
    at Function.fromMergeWithoutPK ({repo_root}/node_modules/@mikro-orm/core/dist/errors.js:30:16)
    at EntityValidator.validatePrimaryKey ({repo_root}/node_modules/@mikro-orm/core/dist/entity/EntityValidator.js:66:44)
    at SqlEntityManager.merge ({repo_root}/node_modules/@mikro-orm/core/dist/EntityManager.js:378:24)
    at SqlEntityManager.map ({repo_root}/node_modules/@mikro-orm/core/dist/EntityManager.js:367:21)
    at {repo_root}/node_modules/@mikro-orm/knex/dist/query/QueryBuilder.js:371:37
    at Array.map (<anonymous>)
    at QueryBuilder.getResultList ({repo_root}/node_modules/@mikro-orm/knex/dist/query/QueryBuilder.js:371:20)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at LightIndicatorService.getStatusInterval ({repo_root}/src/light-indicator/light-indicator.service.ts:358:28)
```

`isEmpty` (`qbStatus.select()`) is returned as `Entity[]` type, that type does not have `isEmpty` property. I suspect that is the reason for this error, but I have no idea how to solve this.

Could you give me a clue please?

**Reply by B4nan** [reply]
You say it works, but works the opposite way as you say. If the table is empty, it will give you no results, if not, it will give you one. So your `isempty=1` means it is not empty, and vice versa, right? 

> Could you give me a clue please?

Don't await the QB, as that means calling `qb.getResult()` which means mapping to entities. Use `qb.execute()` instead.

**Reply by tukusejssirs** [reply]
> You say it works, but works the opposite way as you say. If the table is empty, it will give you no results, if not, it will give you one. So your isempty=1 means it is not empty, and vice versa, right?

Right, the opposite way :man_facepalming:, but works. :wink:

Don't await the QB, as that means calling `qb.getResult()` which means mapping to entities. Use `qb.execute()` instead.

Thanks! :pray:

The following is working. However, I think it might be a good idea to add `isEmpty()` method to check if table is empty.

```ts
const qb = this.orm.em.createQueryBuilder(Entity)
qb.select('1 AS "isEmpty"').limit(1).withSchema(schema)
const result = await qb.execute('get') as {isEmpty?: 1}

if (result?.isEmpty === 1) {
  this.logger.log('Table is not empty.')
} else {
  this.logger.log('Table is empty.')
}
```

**Comment by tukusejssirs** [comment, marked as answer]
As on GitHub Discussions, only top-level comments can be selected as answers, I select this comment as an answer and link to the [answer with example code](https://github.com/mikro-orm/mikro-orm/discussions/2547#discussioncomment-1833702).
